### ISSUE TITLE
Building requires >1GB memory

### ISSUE THREAD

**Issue by TheBlueMatt** [bitcoin/bitcoin#6658]
CXX      test/test_test_bitcoin-script_tests.o
g++: internal compiler error: Killed (program cc1plus)

script_tests requires more than 1GB memory to build, killing a lot of VMs.

**Comment by theuni** [comment]
I don't think an ICE is really our concern. I assume clang manages ok?

**Comment by TheBlueMatt** [comment]
Hmm? I'd say it being impossible to build Bitcoin Core by default on many VPSes is our concern.

**Comment by luke-jr** [comment]
@theuni ICE? The error he posted is from GNU G++...

**Comment by sipa** [comment]
Internal Compiler Errot

**Comment by luke-jr** [comment]
@TheBlueMatt Try using CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768"

**Comment by TheBlueMatt** [comment]
The ICE in question is OOM killer eating g++.

**Comment by TheBlueMatt** [comment]
@Luke-Jr: I can get it to build fine (--disable-tests), I just care that we fix the actual issue.

**Comment by theuni** [comment]
I'd be concerned if no compiler could build on that VPS. However, I'd be willing to bet that it's just a matter of an inefficient gcc version. If clang dies the same way, I'll buy you a beer this weekend :)

**Comment by luke-jr** [comment]
@TheBlueMatt Memory use is not an issue, if it correctly builds with the appropriate CXXFLAGS... IMO

**Comment by TheBlueMatt** [comment]
This is a default gcc on a relatively recent Debian. I really don't think the response to "I can't build Bitcoin Core on a vps with 1GB of memory with default parameters" should be "oh, but if you set these flags you can!"

**Comment by unsystemizer** [comment]
0.11 does require more RAM and in my experience `--disable-tests` doesn't help. 
0.10 could be built on systems with 512MB of RAM (with and without `--disable-tests`).

**Comment by jgarzik** [comment]
Any instrumentation on why script_test takes so much memory?  Would prefer to discern "why" before taking any action.

In general we should be able to build Bitcoin Core on a 1GB VPS.

**Comment by gmaxwell** [comment]
Y'all being overly pedantic.  Matt identified a usability problem: User obtains and attempts to compile the software with normal instructions on a common and reasonable platform type, and it crashes. 

If you want to suggest that the docs should have instructions with low mem cflags or whatnot, OKAY, but this doesn't stop being an issue because us experts can work around it. Our memory usage is kind of obnoxious during compile time IMO, and if we could do some minor tweaks to reduce the peak that would be nice too-- but at the very least we should provide some instructions.

**Comment by laanwj** [comment]
This is actually known and documented: https://github.com/bitcoin/bitcoin/blob/master/doc/build-unix.md#system-requirements

I would have expected that no longer using boost::spirit (for JSON) and boost::asio (for RPC) would reduce compilation memory usage - unfortunate to hear that it became worse. 

Anyhow: cross compilation and pseudo-static compilation is extremely easy with the depends system, I'd recommend using that.

**Comment by laanwj** [comment]
Oh wait, the compilation of that file *uses* more than 1GB of memory?
Ok, reopening, that is definitly not good.

**Comment by TheBlueMatt** [comment]
Yes, the VPS in question has a total of 1.5GB memory (1GB ram + .5 swap), and nothing else of note was running at the time.

**Comment by laanwj** [comment]
I measured using gcc 4.8: Top 10 memory usage during build:

nr | source | elapsed | maxresident (kB)
----- | ----- | ----- | -----
1. |  main.cpp | 	0:36.58 | 	1248524
2. |  init.cpp | 	0:24.24 | 	964136
3. |  wallet/wallet.cpp | 	0:22.11 | 	896096
4. |  test/script_tests.cpp | 	0:19.44 | 	859708
5. |  net.cpp | 	0:18.85 | 	813836
6.  | wallet/rpcwallet.cpp | 	0:14.16 | 	727956
7.  | wallet/walletdb.cpp | 	0:13.48 | 	719388
8. |  wallet/rpcdump.cpp | 	0:12.42 | 	697660
9.  | qt/walletmodel.cpp | 	0:11.26 | 	680500
10.  | rpcrawtransaction.cpp | 	0:11.44 | 	666548

Will try with clang some other time.

**Comment by laanwj** [comment]
Same information, but for clang 3.6.0 (also 64-bit):

nr    | source	 | elapsed	 | maxresident
----- | -------- | ------------- | ----------------------
1 | main.cpp	 | 0:28.34	 | 723960
2 |  wallet/wallet.cpp	 | 0:15.93	 | 522068
3 |  init.cpp	 | 0:19.06	 | 517080
4 |  net.cpp	 | 0:14.36	 | 428924
5 |  wallet/rpcwallet.cpp | 	0:11.53	 | 357224
6 |  wallet/walletdb.cpp | 	0:10.16	 | 356792
7 |  test/script_tests.cpp | 	0:12.73	 | 342376
8 |  qt/walletmodel.cpp | 	0:08.79	 | 333660
9 |  test/test_bitcoin.cpp | 	0:08.21	 | 326100
10 |  wallet/rpcdump.cpp | 	0:09.19	 | 322376

The contenders are the same, although it is faster and uses significantly less memory in all cases. So we should add a recommendation to use clang on lower-memory systems.

**Comment by laanwj** [comment]
I'm not sure why it is this "bad". `main` and `init` have huge complex implementation files, but of `script_tests` you'd say it isn't that bad - just 1016 lines.

Possibly it has to do with embedding the test data as includes:
```
410344  test/data/script_invalid.json.h
530534  test/data/script_valid.json.h
```

**Comment by theuni** [comment]
@laanwj I suspect you're right about the test data. Maybe older gcc re-reads that each time it's needed and balloons in size as a result.

@TheBlueMatt Could you test with https://github.com/theuni/bitcoin/commit/1a107da7ef4b90bff4165357375b58c41f8e01c9 and see if that helps? That moves the test data into cpps instead.

**Comment by TheBlueMatt** [comment]
Hmm, the VPS has since been upgraded to have more memory (yay pruning means I get more memory and less disk space...) and doesnt kill script_tests anymore :( What may have been happening is the host may have been retroactively killing processes after some other process had casued the VM to hit some high memory bar. In any case, it is just Debian Jessie, so gcc 4.9.2-10.

**Comment by TheBlueMatt** [comment]
Still, main max resident of 1.2GB will probably not build on most VPSes today, but sadly thats a harder one to fix.

**Comment by gmaxwell** [comment]
Rather than recommending clang (which would send more hosts into a less tested and potentially consensus inconsistent config) we could recommend compiler flags that reduce memory usage.

**Comment by unsystemizer** [comment]
I ran a comparison test yesterday measuring the peak memory consumption of the Bash shell session in which I was working. gcc - 835 MB, clang - 823 MB (32-bit Raspbian OS). Not a big saving. (I don't know how @laanwj measured RAM consumption, otherwise I would have used his approach).

People low on RAM could use binary builds if they don't have enough RAM to build, or build without tests. Not everyone should be able to build from source on their system (or, in most hosting environments, they can upsize their RAM, build, and then downsize?).

**Comment by laanwj** [comment]
@gmaxwell Possibly both - we're using clang for OSX builds already, so it's not that much of an unknown. I've switched to clang for my run-of-the-mill builds on Linux to save time.

@cfields The solution in https://stackoverflow.com/questions/5479691/is-there-any-standard-way-of-embedding-resources-into-linux-executable-image to use the `.incbin` directive is also quite nice - create a file blob.S:
```
    .global blob
    .global blob_size
    .section .rodata
blob:
    .incbin "blob.bin"
1:
blob_size:
    .int 1b - blob
```
which can just be compiled with `gcc` along with the rest of the source files. This avoids any kind of conversion.

**Comment by jarret** [comment]
After tripping up on this initially, I succeeded in applying Luke-jr's
workaround by giving the CXXFLAGS at configure time:

	$ ./configure --disable-wallet CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768"

On a Raspberry Pi 2 building with a single core:

	$ time make
	(success)
	real    159m41.895s
	user    154m55.090s
	sys     3m1.710s

The memory use seemed to stay below 500 MB throughout, but I didn't measure
that precisely.

**Comment by rebroad** [comment]
Also having trouble compiling main.cpp on a Raspberry Pi. Perhaps this file can be split in two?

**Comment by laanwj** [comment]
@sipa noticed that with recent compilers (gcc 5.3.1 in his case), enabling `-flto` (link time optimization) reduced compile-time memory usage for main.cpp to 850 MiB, without unduly raising link-time memory requirements.
This may be another useful suggestion for limited-memory compiling.

**Comment by TheBlueMatt** [comment]
On gcc 4.9.2, after #9260, objects which fail to build on a kvm host with 1GB of memory (with wallet, no GUI):

libbitcoin_server_a-init.o
libbitcoin_server_a-validation.o
rpc/libbitcoin_server_a-rawtransaction.o
wallet/libbitcoin_wallet_a-rpcdump.o
wallet/libbitcoin_wallet_a-rpcwallet.o
wallet/libbitcoin_wallet_a-wallet.o

**Comment by TheBlueMatt** [comment]
Note that the above list is the same on gcc 6.2.0.

**Comment by laanwj** [comment]
> On gcc 4.9.2, after #9260, objects which fail to build on a kvm host with 1GB of memory (with wallet, no GUI):

This is in contrast to: before #9260?
So splitting up main makes this *worse*? That's not what I would have expected.

**Comment by TheBlueMatt** [comment]
I'm sure its better, but not better enough that both halves dont both end up needing >1GB.

**Comment by gmaxwell** [comment]
I wonder how much less memory it will take once we start eliminating boost in earnest?

**Comment by TheBlueMatt** [comment]
Considering net_processing only took >1GB of memory after I re-added boost/thread.hpp, probably a lot.

**Comment by ciconsulting** [comment]
I am also unable to build on a 1GB VPS running Ubuntu 18.04.

```
Making all in src
make[1]: Entering directory '/home/satoshi/bitcoin-source/bitcoin/src'
make[2]: Entering directory '/home/satoshi/bitcoin-source/bitcoin/src'
make[3]: Entering directory '/home/satoshi/bitcoin-source/bitcoin'
make[3]: Leaving directory '/home/satoshi/bitcoin-source/bitcoin'
  CXX      libbitcoin_server_a-init.o

cc1plus: out of memory allocating 5028920 bytes after a total of 40685568 bytes
Makefile:5606: recipe for target 'libbitcoin_server_a-init.o' failed
make[2]: *** [libbitcoin_server_a-init.o] Error 1
make[2]: Leaving directory '/home/satoshi/bitcoin-source/bitcoin/src'
Makefile:9462: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/home/satoshi/bitcoin-source/bitcoin/src'
Makefile:747: recipe for target 'all-recursive' failed
make: *** [all-recursive] Error 1
```

**Comment by Sjors** [comment]
On a related note, as I'm compiling on a Nano Pi with 1 GB RAM and 4 cores, I initially tried `-j5`, but there are a number of files where that eats into swap and everything grinds to a halt. `-j1` seems the only way to go, but this seems wasteful for the many files that don't require much RAM.

I did use `CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768`. Is there a way to make the compiler use less than e.g. 200 MB RAM per process? Or perhaps it's possible to exclude a specific list of files, run with `-j5` (or `-j3`), and then do those more challenging files with `-j1`?

**Comment by Sjors** [comment]
I'm looking at [Armian builds](https://docs.armbian.com/Developer-Guide_Build-Preparation/), since I might as well just build the rest of the OS - and sync the chain - along with cross compiling bitcoind :-)

**Comment by Rosswestwood** [comment]
Not sure this os the great best way.

**Comment by MarcoFalke** [comment]
Considering that `clang` uses a lot less memory compared to gcc, why don't we mention `clang` as an alternative in https://github.com/bitcoin/bitcoin/blob/000abbb6b07410357a928768d7d56465ba0d3bac/doc/build-unix.md#memory-requirements

Should I create a pull for this?

**Comment by laanwj** [comment]
well if clang, a compiler many, many people use, generates consensus incompatible code, bitcoin has a huge problem.

**Comment by dufferzafar** [comment]
To be honest, I myself use clang (along with ccache) but that's only because all my work involves running in regtest, so I don't have to worry about consensus compatibility.

**Comment by laanwj** [comment]
I mean, it might have been a good argument in 2012, but at this point, many platforms are clang-only (MacOSX, FreeBSD, OpenBSD, at least). There's nothing that binds bitcoin to gcc specifically.

**Comment by meeDamian** [comment]
Just wanted to leave an update here, that building Bitcoin Core v0.16.2 is possible¹ on Raspberry Pi Zero W, which has 512MB RAM + default of 100MB SWAP.

```
./autogen.sh
./configure CXXFLAGS="--param ggc-min-expand=1 --param ggc-min-heapsize=32768" --enable-cxx --without-gui --disable-shared --with-pic
--enable-upnp-default --disable-wallet --disable-tests
make
sudo make install
```

¹ It did take a **very** long time.

**Comment by practicalswift** [comment]
As noted in this issue we are currently using **significantly** more compile-time memory than we technically have to.

I plan to address this and you can see steps being taken in #16273 and #16278 (the latter is a subset of the former limited to tests only).

People who care about compile-time memory usage might be interested in reviewing those PR:s :-)

**Comment by practicalswift** [comment]
These are the 25 worst offenders in terms of compile-time memory usage:

| #  | File | Max memory usage (RSS) |
| -: | :--- | ---------------------: |
| 1 | `test/test_test_bitcoin-script_tests.o` | 1409 MB |
| 2 | `wallet/libbitcoin_wallet_a-wallet.o` | 1102 MB |
| 3 | `wallet/libbitcoin_wallet_a-rpcwallet.o` | 1059 MB |
| 4 | `libbitcoin_server_a-validation.o` | 952 MB |
| 5 | `test/test_test_bitcoin-util_tests.o` | 903 MB |
| 6 | `wallet/test/test_test_bitcoin-wallet_tests.o` | 900 MB |
| 7 | `libbitcoin_server_a-init.o` | 892 MB |
| 8 | `wallet/test/test_test_bitcoin-coinselector_tests.o` | 892 MB |
| 9 | `qt/test/qt_test_test_bitcoin_qt-wallettests.o` | 881 MB |
| 10 | `wallet/test/test_test_bitcoin-ismine_tests.o` | 868 MB |
| 11 | `qt/test/qt_test_test_bitcoin_qt-addressbooktests.o` | 862 MB |
| 12 | `libbitcoin_server_a-net_processing.o` | 857 MB |
| 13 | `wallet/test/test_test_bitcoin-psbt_wallet_tests.o` | 839 MB |
| 14 | `qt/qt_libbitcoinqt_a-sendcoinsdialog.o` | 809 MB |
| 15 | `test/test_test_bitcoin-transaction_tests.o` | 802 MB |
| 16 | `wallet/libbitcoin_wallet_a-rpcdump.o` | 800 MB |
| 17 | `test/test_test_bitcoin-rpc_tests.o` | 782 MB |
| 18 | `libbitcoin_server_a-validationinterface.o` | 771 MB |
| 19 | `rpc/libbitcoin_server_a-blockchain.o` | 744 MB |
| 20 | `qt/qt_libbitcoinqt_a-rpcconsole.o` | 741 MB |
| 21 | `test/test_bitcoin` | 740 MB |
| 22 | `test/test_test_bitcoin-sighash_tests.o` | 732 MB |
| 23 | `rpc/libbitcoin_server_a-rawtransaction.o` | 721 MB |
| 24 | `qt/qt_libbitcoinqt_a-walletmodel.o` | 720 MB |
| 25 | `wallet/libbitcoin_wallet_a-walletdb.o` | 720 MB |
| ... | ... | ... |
| 449 | `compat/libbitcoin_util_a-glibc_sanity.o` | 24 MB |
| 450 | `crypto/sha256_sse4.cpp` | 24 MB |
| 451 | `crypto/crypto_libbitcoin_crypto_base_a-sha256_sse4.o` | 24 MB |
| 452 | `script/libbitcoin_consensus_a-script_error.o` | 23 MB |
| 453 | `script/script_error.cpp` | 23 MB |
| 454 | `compat/libbitcoin_util_a-strnlen.o` | 18 MB |

**Comment by practicalswift** [comment]
People interested in enabling builds on low memory machines might be interested in reviewing and testing #16312: 
* tests: Reduce compile-time memory usage by 0.5 GB when compiling `script_tests.cpp` (from 1.4 GB to 0.9 GB)

**Comment by practicalswift** [comment]
Chasing concept ACK or NACKs for #16331 which tackles this issue.

Let me know if this issue is no longer relevant :-)

**Comment by laanwj** [comment]
It doesn't seem that proposed solutions to this attract much, or any
attention. Most notably, no one building on 1GB menory systems seems to be
helping testing.

So my conclusion would be it's either not relevant anymore or very very low
priority, and will only become less so as systems with so little menory
become rarer.

Sorry for the wasted effort :/

**Comment by meeDamian** [comment]
I'm not sure if I'm not missing something here, but bitcoind was [already building on RBP zero][zero] (512MB ram with just 100MB swap).  Plus, I've just built [v0.18.1] on RBP3 (1GB ram, no swap) a few days ago without problems even before applying any PRs.  It does take its time, however 😅.  So I for once, am not sure what else/how to test?

[zero]: https://github.com/bitcoin/bitcoin/issues/6658#issuecomment-420320388
[v0.18.1]: https://github.com/bitcoin/bitcoin/tree/v0.18.1

**Comment by Sjors** [comment]
@meeDamian does the new PR work without those extra config flags?

**Comment by MarcoFalke** [comment]
It is a bit concerning that the unit test have to be disabled to compile on low memory machines

**Comment by meeDamian** [comment]
On a fresh Raspbian on RBP3, running (with or without `--enable-lowmem` flag):

```
sudo swapoff --all

sudo apt install git build-essential libtool autotools-dev automake pkg-config bsdmainutils python3 libssl-dev libevent-dev libboost-system-dev libboost-filesystem-dev libboost-chrono-dev libboost-test-dev libboost-thread-dev libminiupnpc-dev libzmq3-dev

git clone  -b fix-absurd-memory-usage-when-compiling-script_build  https://github.com/practicalswift/bitcoin.git bitcoin-swift
cd bitcoin-swift/

./contrib/install_db4.sh $(pwd)
export BDB_PREFIX=$(pwd)/db4
./autogen.sh

./configure BDB_LIBS="-L${BDB_PREFIX}/lib -ldb_cxx-4.8" BDB_CFLAGS="-I${BDB_PREFIX}/include" --enable-lowmem --without-gui
```

Crashes with:

```
…
checking for upnpDiscover in -lminiupnpc... yes
checking miniupnpc/miniupnpc.h usability... yes
checking miniupnpc/miniupnpc.h presence... yes
checking for miniupnpc/miniupnpc.h... yes
checking for upnpDiscover in -lminiupnpc... (cached) yes
checking miniupnpc/upnpcommands.h usability... yes
checking miniupnpc/upnpcommands.h presence... yes
checking for miniupnpc/upnpcommands.h... yes
checking for upnpDiscover in -lminiupnpc... (cached) yes
checking miniupnpc/upnperrors.h usability... yes
checking miniupnpc/upnperrors.h presence... yes
checking for miniupnpc/upnperrors.h... yes
checking for upnpDiscover in -lminiupnpc... (cached) yes
checking whether to build Bitcoin Core GUI... no (Qt5)
checking for boostlib >= 1.47.0 (104700)... yes
checking whether the Boost::System library is available... yes
configure: error: Could not find a version of the library!
```

Running the same on bitcoin `v0.18.1` does not crash.

**Comment by meeDamian** [comment]
Actually… bitcoin `master` also crashes the same way, so likely unrelated to this.  Happy to try again if someone advises how to fix/proceed.

**Comment by Sjors** [comment]
@meeDamian did you try with `practicalswift/low-mem`? #16331

**Comment by meeDamian** [comment]
I'm afraid it crashes the same way as `master`, and `fix-absurd-memory-usage-when-compiling-script_build`.  Any idea what library version might be missing there, or how to fix it?

**Comment by MarcoFalke** [comment]
The docs are updated and mention that `clang` can be used, which uses less than 1 GB of memory. Closing for now

https://github.com/bitcoin/bitcoin/blob/561a7d30478b82f5d46dcf0f16e864a9608004f4/doc/build-unix.md#memory-requirements
